### PR TITLE
Support TypeScript `.js` imports

### PR DESCRIPTION
**Pull request by heyheyhello**
Closes https://github.com/rollup/plugins/issues/479

## Rollup Plugin Name: `node-resolve`

This PR contains:

- [x] bugfix
- [x] feature
- [ ] refactor
- [ ] documentation
- [ ] other

Are tests included?

- [x] yes (_bugfixes and features will not be merged without tests_)
- [ ] no

Breaking Changes?

- [ ] yes (_breaking changes will not be merged unless absolutely necessary_)
- [x] no

If yes, then include "BREAKING CHANGES:" in the first commit message body, followed by a description of what is breaking. 

List any relevant issue numbers:

https://github.com/rollup/plugins/issues/479

### Description

As described in #479 it's common for a TS project to use '.js' to refer to local '.ts' or '.tsx' files. This adds support for that.

Previously/Currently you need to import as either `'./file'` (see quote in #479 :wink:) or `'./file.ts'`. Both don't really work:

image

image

The fix:

image

I haven't added tests because it's only 5 lines of code and the "algorithm" is just "Is a TS/TSX file importing a JS file and `{ extensions: [] }` has TS and/or TSX in it? Then add to `importSpecifierList` as a potential resolution".

Note that `@rollup/plugin-typescript` works out of the box for this already, but I think it'd be good to not _need_ to require the full TS compiler just to support this - VSCode already lints the whole project, I don't need a compile, I just need types removed via Babel (so so much faster)

### CODE REVIEW

**Comment by shellscape** [comment]
Thanks for the PR. Unfortunately tests for all fixes a s features are a hard requirement. We won't be able to accept your PR until tests for this are added.

**Comment by heyheyhello** [comment]
All tests pass on my system:

image
image


@shellscape Could let know what test is failing? Circle CI won't let me see their test output without giving them write access to all my repos which is pretty disgusting - so I won't be doing that. I checked the "Contributing" section of your README.md but apparently there's an extra step I'm missing.

Thanks!

**Comment by NotWoods** [comment]
Looks like a linting error is causing the CI failure:

```
> @rollup/plugin-node-resolve@8.1.0 lint:js /home/circleci/project/packages/node-resolve
> eslint --fix --cache src test types --ext .js,.ts
 
/home/circleci/project/packages/node-resolve/test/fixtures/extensions/main.js
  0:0  error  Parsing error: "parserOptions.project" has been set for @typescript-eslint/parser.
The file does not match your project config: packages/node-resolve/test/fixtures/extensions/main.js.
The file must be included in at least one of the projects provided

 
✖ 1 problem (1 error, 0 warnings)
 
```

**Comment by heyheyhello** [comment]
Great thanks for the quick response! I now have it happening on my side too. Looking into it

**Comment by heyheyhello** [comment]
The last commit has a bunch of fixes that aren't related to #479. Happens 🤷

**Comment by heyheyhello** [comment]
@shellscape (or anyone) is there anything holding up this PR from being merged? If the concept is controversial or something let me know.

**Comment by heyheyhello** [comment]
Ok sounds good, I just wanted to check it was live. No rush. Thanks!

File: .eslintrc.js [hunk 1]
```diff
@@ -1,25 +1,35 @@
 module.exports = {
-  extends: ['rollup', 'plugin:import/typescript'],
+  extends: ['rollup'],
```
Comment (shellscape) [inline .eslintrc.js:2]: why is this change necessary? I'm using virtually the same config in several projects without issue.

**Comment by heyheyhello** [comment]
For "conventional commits" are you asking me to squash the work into one commit called `fix(node-resolve): support .js extension #479`? I know your repo uses conventional commits but really it feels like it should be up to you as a merger to decide what commits you want. If this a fix or a feat?

> PRs which touch multiple plugins are generally discouraged. For those that must be, each plugin should have it's own individual commits. Commits shouldn't touch more than one plugin

This PR only wants to touch `node-resolve`. The very last commit that runs ESLint to autofix problems that _somehow_ snuck into your repo to fix semicolons/quotes isn't part of the PR. If you don't want your own ESLint fixes that's fine - I'll remove the commit and you can run the linter yourself and merge it into master. I'm not opening 4-5 PRs for a semicolon fix...

I'll rebase with a squashed commit following your naming convention and remove the commit that does ESLint fixes.

**Comment by shellscape** [comment]
> I'll rebase with a squashed commit following your naming convention and remove the commit that does ESLint fixes.

Good deal. I addressed all of the changed files so as not to minimize any work that went into them.

File: .eslintrc.js (hunk 1 above)
Comment (heyheyhello) [reply .eslintrc.js:2]: I factored out all TS-related ESLint stuff to one "override" block that only operates on .ts/.tsx. So the "extends" has import/typescript in there instead of the general JS config

Comment (heyheyhello) [reply .eslintrc.js:2]: If you want it in both JS and TS I could set that, but it feels like it belongs in only TS?

**Comment by heyheyhello** [comment]
Hmm @shellscape do you know anything about the broken lockfile? There was a conflict so I deleted the lockfile and reran `pnpm i`. CircleCI is fine with it but your GitHub checks fail :/

**Comment by shellscape** [comment]
Give `pnpm install --no-frozen-lockfile` a try. If that doesn't resolve it I'll make some time to fix it.

**Comment by heyheyhello** [comment]
Oh yeah I should have mentioned I did try that. I'll keep poking around

**Comment by heyheyhello** [comment]
It seems like you're working on dynamic-import-vars right now which might be related...

```
ERROR  Cannot install with "frozen-lockfile" because pnpm-lock.yaml is not up-to-date with packages\dynamic-import-vars\package.json
```

**Comment by shellscape** [comment]
Might have to merge from upstream/master maybe.

**Comment by heyheyhello** [comment]
✨✨✨

Ok @shellscape if you can reply to the thread/review you opened that's all

File: .eslintrc.js (hunk 1 above)
Comment (shellscape) [reply .eslintrc.js:2]: I'd prefer to revert this change and take a look at the merits of modifying it in an issue, followed by a repo-level commit if we want to move in that direction. The reason being is that this will effect all open PRs and development for all plugins moving forward, and I'm not ready to commit to that big a change.

Comment (heyheyhello) [reply .eslintrc.js]: Yeah no worries. It just seemed to make sense at the time but I have no preference. Reverted the change

Comment (shellscape) [reply .eslintrc.js]: my apologies if it was unclear, but I was speaking to all of the changes in the file. We're totally open to discussing a refactor of the config in another pipeline.

Comment (heyheyhello) [reply .eslintrc.js]: Oh - that was not clear yeah.

I brought these changes in because the build fails without them.

From @NotWoods 

> Looks like a linting error is causing the CI failure:
> 
> ```
> > @rollup/plugin-node-resolve@8.1.0 lint:js /home/circleci/project/packages/node-resolve
> > eslint --fix --cache src test types --ext .js,.ts
>  
> /home/circleci/project/packages/node-resolve/test/fixtures/extensions/main.js
>   0:0  error  Parsing error: "parserOptions.project" has been set for @typescript-eslint/parser.
> The file does not match your project config: packages/node-resolve/test/fixtures/extensions/main.js.
> The file must be included in at least one of the projects provided
> 
>  
> ✖ 1 problem (1 error, 0 warnings)
>  
> ```

This is how I limit the TS parser to only TS files (https://github.com/typescript-eslint/typescript-eslint/issues/1928). I'll look into doing it another way tomorrow

Comment (shellscape) [reply .eslintrc.js]: I see, sorry that I missed that. You can place an overriding `.eslintrc.js` file in that test directory to modify that behavior.  we've done that often in the repo and is perfectly acceptable. If there are too many errors to contend with, you can add that to the `.eslintignore` file. Fixtures always get wiggle room.

**Comment by heyheyhello** [comment]
> why is this change necessary? I'm using virtually the same config in several projects without issue.

That's seems odd because I actually have to commit with `git commit --no-verify` to bypass ESLint failing on `.eslinrc.js` with the old config: 

image

But if it works for you that's what's important. Anyway I fixed it (on my end; we'll see if GitHub likes it) by moving TS files to their own folder - it seems ESLint will operate on entire folders so any JS files will be sucked into the TS mess.

**Comment by heyheyhello** [comment]
Oops.

**Comment by shellscape** [comment]
OK, I'll clone your fork and branch to take a look.

**Comment by heyheyhello** [comment]
My bad about the tests

**Comment by heyheyhello** [comment]
:star: Looks good

**Comment by shellscape** [comment]
Thanks for working with us on this one.

**Comment by heyheyhello** [comment]
Thanks both of you for helping to get it merged in :)
